src/AutoPilotPlugins/APM/APMAirframeComponentController.h

```diff
@@ -0,0 +1,140 @@
+#include "AutoPilotPlugin.h"
+#include "FactPanelController.h"
+#include "APMAirframeComponentAirframes.h"
+
+class APMAirframeModel;
+
+/// MVC Controller for APMAirframeComponent.qml.
+class APMAirframeComponentController : public FactPanelController
+{
+    Q_OBJECT
+    
+public:
+    APMAirframeComponentController(void);
+    ~APMAirframeComponentController();
+    
+    Q_PROPERTY(bool showCustomConfigPanel MEMBER _showCustomConfigPanel NOTIFY showCustomConfigPanelChanged)
+    Q_PROPERTY(APMAirframeModel* airframeTypesModel MEMBER _airframeTypesModel CONSTANT)
+    Q_PROPERTY(QString currentAirframeType MEMBER _currentAirframeType CONSTANT)
+    Q_PROPERTY(QString currentVehicleName MEMBER _currentVehicleName CONSTANT)
+    Q_PROPERTY(int currentVehicleIndex MEMBER _currentVehicleIndex CONSTANT)
+    Q_PROPERTY(int autostartId MEMBER _autostartId NOTIFY autostartIdChanged)
+    Q_PROPERTY(QString fileParams MEMBER _currentFileParams)
+
+    Q_INVOKABLE void changeAutostart(void);
+
+    
+    int currentAirframeIndex(void);
+    void setCurrentAirframeIndex(int newIndex);
+    
+signals:
+    void loadAirframesCompleted();
+    void autostartIdChanged(int newAutostartId);
+    void showCustomConfigPanelChanged(bool show);
+    
+private slots:
+    void _waitParamWriteSignal(QVariant value);
+    void _rebootAfterStackUnwind(void);
+    void _fillAirFrames(void);
+private:
+    static bool _typesRegistered;
+    QString         _currentAirframeType;
+    QString         _currentVehicleName;
+    int             _currentVehicleIndex;
+    int             _autostartId;
+    QString         _currentFileParams;
+    bool            _showCustomConfigPanel;
+    int             _waitParamWriteSignalCount;
+    APMAirframeModel    *_airframeTypesModel;
+};
+
+class APMAirframe : public QObject
+{
+    Q_OBJECT
+    
+public:
+    APMAirframe(const QString& name, const QString& paramsFile, int autostartId, QObject* parent = NULL);
+    ~APMAirframe();
+    
+    Q_PROPERTY(QString text MEMBER _name CONSTANT)
+    Q_PROPERTY(int autostartId MEMBER _autostartId CONSTANT)
+    Q_PROPERTY(QString params MEMBER _paramsFile CONSTANT)
+    
+private:
+    QString _name;
+    QString _paramsFile;
+    int     _autostartId;
+};
+
+class APMAirframeType : public QObject
+{
+    Q_OBJECT
+    
+public:
+    APMAirframeType(const QString& name, const QString& imageResource, QObject* parent = NULL);
+    ~APMAirframeType();
+    
+    Q_PROPERTY(QString name MEMBER _name CONSTANT)
+    Q_PROPERTY(QString imageResource MEMBER _imageResource CONSTANT)
+    Q_PROPERTY(QVariantList airframes MEMBER _airframes CONSTANT)
+    
+    void addAirframe(const QString& name, const QString& paramsFile, int autostartId);
+    QString name() const;
+    QString imageResource() const;
+
+private:
+    QString         _name;
+    QString         _imageResource;
+    QVariantList    _airframes;
+};
+
+class APMAirframeModel : public QAbstractListModel {
```

Comment:
You don't need to build one of these yourself. QmlObjectListModel.cc/h alraedy exists which supports creating a QmlObjectListModel from any QObject.